### PR TITLE
Enhance type error understanding by displaying incompatible constraints

### PR DESCRIPTION
**Pull request by Gueckmooh**
Sometimes it is complicated to understand errors such as `Unable to deduce type for variable a`. I tried to add more information on the cause of the error.

For now it is still a work in progress, but I have a few results.

## Example

```souffle
.decl foo(a: number, b: number)
.decl foo(a: symbol, b: number)

foo(1, 2).
bar(a, b) :-
   foo(a, b).
```

### Before

```
$ souffle test.dl
Error: Unable to deduce type for variable a in file test.dl at line 5
bar(a, b) :-
----^--------
Error: Unable to deduce type for variable a in file test.dl at line 6
    foo(a, b).
--------^------
2 errors generated, evaluation aborted
```

### After

```
$ souffle test.dl
Error: Unable to deduce type for variable a in file test.dl at line 5
bar(a, b) :-
----^--------
Following constraints are incompatible:
   Type of variable 'a' should be a subtype of 'symbol'
   Type of variable 'a' should be a subtype of 'number' in file test.dl at line 6
    foo(a, b).
----^----------
1 errors generated, evaluation aborted
```

## How

I added an `ErrorAnalyzer` class that collects the minimal set of incompatible constraints on a problem, similarly to the unsat core of a SAT problem. It is injected into the type analysis to gather those information.

Then when emitting errors, it is used to "explain" why the error occurs.

## To do

- [x] Deduplicate errors (merge line 5 and line 6 errors in the example)
- [x] Associate source location information to constraints
- [x] Get user friendly messages to explain failing constraints
- [x] Activate the "error understanding" mechanism on option
- [x] Unit tests
- [x] Remove the option
- [x] Prettify constant type names

~~Make deduplication less ugly~~ it's ugly but only when an error occurs

Let me know if the change looks too intrusive.

### CODE REVIEW

File: src/ast/analysis/typesystem/TypeConstraints.cpp [hunk 1]
```diff
@@ -30,6 +30,7 @@ TypeConstraint isSubtypeOf(const TypeVar& a, const TypeVar& b) {
  * a are subtypes of type b.
  */
 TypeConstraint isSubtypeOf(const TypeVar& variable, const Type& type) {
+    // @here
```
Comment (quentin) [inline src/ast/analysis/typesystem/TypeConstraints.cpp:33]: cleanup?

File: src/ast/analysis/ValueChecker.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file ErrorAnalyzer.h
+ *
+ * Extract reasons why something went wrong.
```
Comment (quentin) [inline src/ast/analysis/ValueChecker.h:13]: copy/pasted

File: src/ast/analysis/ConstraintSystem.h [hunk 3]
```diff
@@ -409,6 +430,30 @@ class Problem {
         return assignment;
     }
 
+    unsat_core_type extractUnsatCore(  // const problem_type& constraints,
```
Comment (quentin) [inline src/ast/analysis/ConstraintSystem.h:433]: to cleanup

File: src/ast/analysis/ValueChecker.h (hunk 2 above)
Comment (Gueckmooh) [reply src/ast/analysis/ValueChecker.h:13]: Woops

File: tests/semantic/agg_checks/agg_checks.err [hunk 4]
```diff
@@ -10,33 +10,26 @@ r("X",Y) :- Y = min X : { X != Y }.
 Error: Unable to deduce type for variable Y in file agg_checks.dl at line 11
 r("X",Y) :- Y = min X : { X != Y, !a(X,Y) }.
 ------^--------------------------------------
-Error: Unable to deduce type for variable Y in file agg_checks.dl at line 11
-r("X",Y) :- Y = min X : { X != Y, !a(X,Y) }.
-------------^--------------------------------
+Following constraints are incompatible:
```
Comment (quentin) [inline tests/semantic/agg_checks/agg_checks.err:13]: No constraint listed here ?

**Review by julienhenry — APPROVED** [review]

**Pull request merged**